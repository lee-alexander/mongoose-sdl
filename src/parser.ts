import { promises as fs } from 'fs';
import { groupItemsBy, toDictionary, uniqueValues, unwrap } from './util';
import { Enum, Schema, Document, DbDefinition, SchemaDataTypeWithoutArray } from './types';

// Break overall document down into key sections - enums, documents, schemas
const TopLevelRegex = /(enum|document|schema) ([^{]*) {([^}]*)}/g;

export async function parseDbDefinitionFile(path: string): Promise<DbDefinition> {
  const fileBuffer = await fs.readFile(path);
  const contents = fileBuffer.toString();
  const parsedContent = [...contents.matchAll(TopLevelRegex)];

  const namesWithType = parsedContent.map(([, type, name]) => `${name}${type}`);
  if (namesWithType.length !== uniqueValues(namesWithType).length) {
    throw new Error('Duplicate type name detected');
  }

  const contentByType = groupItemsBy(
    parsedContent,
    ([, type]) => type as 'enum' | 'document' | 'schema',
    ([, , name, contents]) => ({ name, contents })
  );

  const namedTypes: NamedTypes = {
    enums: new Set((contentByType['enum'] ?? []).map((d) => d.name)),
    documents: new Set((contentByType['document'] ?? []).map((d) => d.name)),
    schemas: new Set((contentByType['schema'] ?? []).map((d) => d.name)),
  };

  const enums = toDictionary(
    contentByType['enum'] ?? [],
    ({ name }) => name,
    ({ contents }) => parseEnumContents(contents)
  );

  const schemas = toDictionary(
    contentByType['schema'] ?? [],
    ({ name }) => name,
    ({ contents }) => parseSchemaContents(contents, namedTypes)
  );

  const documents = toDictionary(
    contentByType['document'] ?? [],
    ({ name }) => name,
    ({ contents }) => parseDocumentContents(contents, namedTypes)
  );

  return { enums, schemas, documents };
}

function parseEnumContents(contents: string): Enum {
  return {
    values: contents.trim().split(/\s+/g),
  };
}

// Break down "key: type @directive1 @directive2" lines into
// (1) key
// (2) array element type, if array
// (3) array element type is required (!), if array
// (4) type, if not array
// (5) type is required (!)
// (6) directives
const KeyValueDirectiveRegex = /(\w+): (?:\[(\w+)(!?)\]|(\w+))(!?)([^\n]*)/g;

// Break down @directive1 @directive2 into (1) directive1 (2) directive2
const DirectiveRegex = /@(\w+)/g;

function parseSchemaContents(contents: string, namedTypes: NamedTypes): Schema {
  const parsedContent = [...contents.matchAll(KeyValueDirectiveRegex)];
  const parsedFields = parsedContent.map(
    ([, fieldName, arrayElementType, arrayElementRequired, nonArrayType, required, directives]) => ({
      fieldName,
      fieldType: unwrap(nonArrayType || arrayElementType),
      isArray: Boolean(arrayElementType),
      isArrayElementRequired: arrayElementRequired === '!',
      isRequired: required === '!',
      directives: directives ? [...directives.matchAll(DirectiveRegex)].map(([, directive]) => directive) : [],
    })
  );

  return toDictionary(
    parsedFields,
    (f) => f.fieldName,
    (f) => ({
      dataType: f.isArray
        ? {
            type: 'Array' as const,
            elementType: parseDataType(f.fieldName, f.fieldType, namedTypes),
            elementRequired: f.isArrayElementRequired,
          }
        : parseDataType(f.fieldName, f.fieldType, namedTypes),
      isRequired: f.isRequired,
      isIndex: f.directives.includes('index'),
      isUnique: f.directives.includes('unique'),
    })
  );
}

function parseDocumentContents(contents: string, namedTypes: NamedTypes): Document {
  return {
    schema: parseSchemaContents(contents, namedTypes),
  };
}

function parseDataType(fieldName: string, fieldType: string, namedTypes: NamedTypes): SchemaDataTypeWithoutArray {
  if (fieldType === 'String' || fieldType === 'Boolean' || fieldType === 'Number') {
    return { type: fieldType };
  }

  if (namedTypes.enums.has(fieldType)) {
    return {
      type: 'Enum',
      refEnum: fieldType,
    };
  }

  if (namedTypes.documents.has(fieldType)) {
    return {
      type: 'ObjectId',
      refDocument: fieldType,
    };
  }

  if (namedTypes.schemas.has(fieldType)) {
    return {
      type: 'Schema',
      refSchema: fieldType,
    };
  }

  throw new Error(`Field ${fieldName} has unknown type ${fieldType}`);
}

interface NamedTypes {
  enums: Set<string>;
  schemas: Set<string>;
  documents: Set<string>;
}
