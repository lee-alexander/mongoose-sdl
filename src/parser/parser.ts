import { promises as fs } from 'fs';
import { groupItemsBy, toDictionary, uniqueValues, unwrap } from '../util';
import { Enum, Schema, Model, DbDefinition, FlatSchemaDataType, Union, Unions } from '../types';

// Break overall document down into key sections - enums, models, schemas
const TopLevelRegex = /(?:(enum|model|schema) ([^{]*) {([^}]*)})|(?:external (\w*))|(?:union (\w*) = ([\w |]*))/g;

export async function parseDbDefinitionFile(path: string): Promise<DbDefinition> {
  const fileBuffer = await fs.readFile(path);
  const contents = fileBuffer.toString();
  const parsedContent = parseContent(contents, TopLevelRegex);

  const topLevelNames = parsedContent.map(
    ([, , name, _, externalTypeName, unionName]) => name || externalTypeName || unionName
  );
  if (topLevelNames.length !== uniqueValues(topLevelNames).length) {
    throw new Error('Duplicate schema/model/enum/external/union name detected');
  }

  const contentByType = groupItemsBy(
    parsedContent,
    ([, type, , , externalTypeName, unionName]) =>
      externalTypeName ? 'external' : unionName ? 'union' : (type as 'enum' | 'model' | 'schema'),
    ([, , name, contents, externalTypeName, unionName, unionContents]) => ({
      name: name || externalTypeName || unionName,
      contents: contents || unionContents || '',
    })
  );

  const namedTypes: NamedTypes = {
    enums: new Set((contentByType['enum'] ?? []).map((d) => d.name)),
    models: new Set((contentByType['model'] ?? []).map((d) => d.name)),
    schemas: new Set((contentByType['schema'] ?? []).map((d) => d.name)),
    externals: new Set((contentByType['external'] ?? []).map((d) => d.name)),
    unions: new Set((contentByType['union'] ?? []).map((d) => d.name)),
  };

  const unions = toDictionary(
    contentByType['union'] ?? [],
    ({ name }) => name,
    ({ name, contents }) => parseUnionContents(name, contents, namedTypes)
  );

  const enums = toDictionary(
    contentByType['enum'] ?? [],
    ({ name }) => name,
    ({ contents }) => parseEnumContents(contents)
  );

  const schemas = toDictionary(
    contentByType['schema'] ?? [],
    ({ name }) => name,
    ({ contents }) => parseSchemaContents(contents, namedTypes, unions)
  );

  const models = toDictionary(
    contentByType['model'] ?? [],
    ({ name }) => name,
    ({ contents }) => parseModelContents(contents, namedTypes, unions)
  );

  const externals = (contentByType['external'] ?? []).map((c) => c.name);

  return { enums, schemas, models, unions, externals };
}

const EnumRegex = /(\w+)/g;
function parseEnumContents(contents: string): Enum {
  const values = parseContent(contents, EnumRegex);

  return {
    values: values.map(([value]) => value),
  };
}

function parseUnionContents(name: string, contents: string, namedTypes: NamedTypes): Union {
  const refModels = contents.split('|').map((m) => m.trim());
  const unknownModels = refModels.filter((model) => !namedTypes.models.has(model));
  if (unknownModels.length > 0) {
    throw new Error(`Unknown model names in union ${name}: ${unknownModels.join(', ')}`);
  }
  return { refModels };
}

// Break down "key: type @directive1 @directive2" lines into
// (1) key
// (2) array element type, if array
// (3) array element type is required (!), if array
// (4) map element type, if map
// (5) map element type is required (!), if map
// (6) type, if not array or map
// (7) type is required (!)
// (8) directives
const KeyValueDirectiveRegex = /(\w+): (?:\[(\w+)(!?)\]|Map<(\w+)(!?)>|(\w+))(!?)([^\n]*)/g;

// Break down @directive1 @directive2 into (1) directive1 (2) directive2
const DirectiveRegex = /@(\w+)/g;

function parseSchemaContents(contents: string, namedTypes: NamedTypes, unions: Unions): Schema {
  const parsedContent = parseContent(contents, KeyValueDirectiveRegex);
  const parsedFields = parsedContent.map(
    ([
      ,
      fieldName,
      arrayElementType,
      arrayElementRequired,
      mapElementType,
      mapElementRequired,
      nonArrayType,
      required,
      directives,
    ]) => ({
      fieldName,
      fieldType: unwrap(nonArrayType || arrayElementType || mapElementType),
      isArray: Boolean(arrayElementType),
      isArrayElementRequired: arrayElementRequired === '!',
      isMap: Boolean(mapElementType),
      isMapElementRequired: mapElementRequired === '!',
      isRequired: required === '!',
      directives: directives ? parseContent(directives, DirectiveRegex).map(([, directive]) => directive) : [],
    })
  );

  const invalidDirectives = parsedFields
    .flatMap((f) => f.directives)
    .filter((d) => d !== 'index' && d !== 'unique' && d !== 'immutable' && d !== 'virtual' && d !== 'validate');
  if (invalidDirectives.length > 0) {
    throw new Error('Unknown directives: ' + invalidDirectives);
  }

  return toDictionary(
    parsedFields,
    (f) => f.fieldName,
    (f) => ({
      dataType: f.isArray
        ? {
            type: 'Array' as const,
            elementType: parseDataType(f.fieldName, f.fieldType, namedTypes, unions),
            elementRequired: f.isArrayElementRequired,
          }
        : f.isMap
        ? {
            type: 'Map' as const,
            elementType: parseDataType(f.fieldName, f.fieldType, namedTypes, unions),
            elementRequired: f.isMapElementRequired,
          }
        : parseDataType(f.fieldName, f.fieldType, namedTypes, unions),
      isRequired: f.isRequired,
      isIndex: f.directives.includes('index'),
      isUnique: f.directives.includes('unique'),
      isImmutable: f.directives.includes('immutable'),
      isVirtual: f.directives.includes('virtual'),
      isValidatable: f.directives.includes('validate'),
    })
  );
}

function parseModelContents(contents: string, namedTypes: NamedTypes, unions: Unions): Model {
  return {
    schema: parseSchemaContents(contents, namedTypes, unions),
  };
}

function parseDataType(
  fieldName: string,
  fieldType: string,
  namedTypes: NamedTypes,
  unions: Unions
): FlatSchemaDataType {
  if (fieldType === 'String' || fieldType === 'Boolean' || fieldType === 'Number' || fieldType === 'Date') {
    return { type: fieldType };
  }

  if (namedTypes.enums.has(fieldType)) {
    return {
      type: 'Enum',
      refEnum: fieldType,
    };
  }

  if (fieldType === 'ObjectId') {
    return {
      type: 'ObjectId',
      refModels: [],
    };
  }

  if (namedTypes.externals.has(fieldType)) {
    return {
      type: 'External',
      refType: fieldType,
    };
  }

  if (namedTypes.models.has(fieldType)) {
    return {
      type: 'ObjectId',
      refModels: [fieldType],
    };
  }

  if (unions[fieldType]) {
    return {
      type: 'ObjectId',
      refModels: unions[fieldType].refModels,
    };
  }

  if (namedTypes.schemas.has(fieldType)) {
    return {
      type: 'Schema',
      refSchema: fieldType,
    };
  }

  throw new Error(`Field ${fieldName} has unknown type ${fieldType}`);
}

interface NamedTypes {
  enums: Set<string>;
  schemas: Set<string>;
  models: Set<string>;
  externals: Set<string>;
  unions: Set<string>;
}

function parseContent(content: string, regex: RegExp) {
  const result = content.matchAll(regex);
  const unmatched = content.replaceAll(regex, '').trim();
  if (unmatched.length > 0) {
    throw new Error('Unexpected syntax near:\n' + unmatched);
  }
  return [...result];
}
