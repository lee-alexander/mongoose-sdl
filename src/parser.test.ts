import { parseDbDefinitionFile } from './parser';

test('parse example.mgsdl', async () => {
  const result = await parseDbDefinitionFile('./sample-data/example.mgsdl');
  expect(result).toMatchSnapshot();
});
