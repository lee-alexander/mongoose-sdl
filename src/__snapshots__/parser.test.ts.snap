// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse example.mgsdl 1`] = `
{
  "documents": {
    "Project": {
      "schema": {
        "creator": {
          "dataType": {
            "refDocument": "User",
            "type": "ObjectId",
          },
          "isIndex": true,
          "isRequired": true,
          "isUnique": false,
        },
        "details": {
          "dataType": {
            "refSchema": "ProjectDetails",
            "type": "Schema",
          },
          "isIndex": false,
          "isRequired": true,
          "isUnique": false,
        },
      },
    },
    "User": {
      "schema": {
        "email": {
          "dataType": {
            "type": "String",
          },
          "isIndex": true,
          "isRequired": true,
          "isUnique": true,
        },
        "name": {
          "dataType": {
            "type": "String",
          },
          "isIndex": false,
          "isRequired": false,
          "isUnique": false,
        },
        "status": {
          "dataType": {
            "refEnum": "UserStatus",
            "type": "Enum",
          },
          "isIndex": false,
          "isRequired": true,
          "isUnique": false,
        },
      },
    },
  },
  "enums": {
    "UserStatus": {
      "values": [
        "Pending",
        "Active",
        "Inactive",
      ],
    },
  },
  "schemas": {
    "ProjectDetails": {
      "alternateNames": {
        "dataType": {
          "elementRequired": true,
          "elementType": {
            "type": "String",
          },
          "type": "Array",
        },
        "isIndex": false,
        "isRequired": true,
        "isUnique": false,
      },
      "isActive": {
        "dataType": {
          "type": "Boolean",
        },
        "isIndex": false,
        "isRequired": true,
        "isUnique": false,
      },
      "name": {
        "dataType": {
          "type": "String",
        },
        "isIndex": false,
        "isRequired": true,
        "isUnique": false,
      },
      "number": {
        "dataType": {
          "type": "Number",
        },
        "isIndex": false,
        "isRequired": false,
        "isUnique": true,
      },
    },
  },
}
`;
